source/common/nighthawk_service_client_impl.cc

```diff
@@ -0,0 +1,42 @@
+#include "common/nighthawk_service_client_impl.h"
+
+#include "external/envoy/source/common/common/assert.h"
+
+namespace Nighthawk {
+
+absl::StatusOr<nighthawk::client::ExecutionResponse>
+NighthawkServiceClientImpl::PerformNighthawkBenchmark(
+    nighthawk::client::NighthawkService::StubInterface* nighthawk_service_stub,
+    const nighthawk::client::CommandLineOptions& command_line_options) {
+  nighthawk::client::ExecutionRequest request;
+  nighthawk::client::ExecutionResponse response;
+  *request.mutable_start_request()->mutable_options() = command_line_options;
+
+  ::grpc::ClientContext context;
+  std::shared_ptr<::grpc::ClientReaderWriterInterface<nighthawk::client::ExecutionRequest,
+                                                      nighthawk::client::ExecutionResponse>>
+      stream(nighthawk_service_stub->ExecutionStream(&context));
+
+  if (!stream->Write(request)) {
+    return absl::UnknownError("Failed to write request to the Nighthawk Service gRPC channel.");
+  } else if (!stream->WritesDone()) {
+    return absl::UnknownError("WritesDone() failed on the Nighthawk Service gRPC channel.");
+  }
+
+  bool got_response = false;
+  while (stream->Read(&response)) {
+    RELEASE_ASSERT(!got_response,
+                   "Nighthawk Service has started responding with more than one message.");
+    got_response = true;
+  }
+  if (!got_response) {
+    return absl::UnknownError("Nighthawk Service did not send a gRPC response.");
```

Comment:
This one doesn't seem like an Unknown Error.  Can we say this is an internal error?  As currently, this is something that shouldn't happen and is entirely internal to the codebase

Reply:
After looking at the docs, I found there is Unavailable, meaning you to retry the request as-is. This might apply if the original Write fails since the Nighthawk Service might just have been down. But if you managed to Write but then got an error from WritesDone, Read, or Finish, those could all be Internal since they shouldn't happen. Does that make sense?